### PR TITLE
small builder updates

### PR DESCRIPTION
**Pull request by lumag**
I started rebasing my code to use `CertificateBuilder` and I had to make several useful modifications.

Note: It is still not possible to create RSA PSS certificates, since there is no `Signer` implementation for RSA PSS, only `RandomizedSigner` (there is no standard detereministic RSA PSS signature process).  Locally I have changed the `CertificateBuilder` to accept `RandomizedSigner` instead of bare `Signer`, however this doesn't look like an universal solution too.

### CODE REVIEW

**Comment by lumag** [comment]
cc @baloo @tarcieri

**Comment by lumag** [comment]
@baloo another question: would it be useful to accept `PK: EncodePublicKey` instead of `SPKIOwned`?

File: x509-cert/src/builder.rs [hunk 1]
```diff
@@ -223,35 +225,32 @@ impl Profile {
...
 where
-    S: Keypair,
+    S: Keypair + DynSignatureAlgorithmIdentifier,
     S::VerifyingKey: EncodePublicKey,
-    S::VerifyingKey: DynSignatureAlgorithmIdentifier,
 {
     /// Creates a new certificate builder
     pub fn new<Signature>(
         profile: Profile,
-        version: Version,
```
Comment (baloo) [inline x509-cert/src/builder.rs:249]: good call. Thank you!

File: x509-cert/src/builder.rs [hunk 2]
```diff
@@ -161,20 +166,18 @@ impl Profile {
...
         match self {
             Profile::Root | Profile::SubCA { .. } => {
-                extensions.push(
-                    KeyUsage(
-                        KeyUsages::DigitalSignature | KeyUsages::KeyCertSign | KeyUsages::CRLSign,
-                    )
-                    .to_extension(tbs)?,
-                );
+                extensions
+                    .push(KeyUsage(KeyUsages::KeyCertSign | KeyUsages::CRLSign).to_extension(tbs)?);
```
Comment (baloo) [inline x509-cert/src/builder.rs:170]: `KeyUsages::DigitalSignature` is getting dropped, I don't remember why I added it in the first place.

**Comment by tarcieri** [comment]
>  It is still not possible to create RSA PSS certificates, since there is no `Signer` implementation for RSA PSS, only `RandomizedSigner` (there is no standard detereministic RSA PSS signature process).  Locally I have changed the `CertificateBuilder` to accept `RandomizedSigner` instead of bare `Signer`, however this doesn't look like an universal solution too.

A couple potential options here:

1. Move the signer out of `CertificateBuilder::new` (which would get rid of the lifetime on the builder), replacing the `build` method with `sign` and `sign_with_rng` methods, which could accept either a `Signer` or `RandomizedSigner`
2. Add a `getrandom` feature to `rsa` and use it to gate a `Signer` impl on `pss::SigningKey` which uses `OsRng` for randomness

File: x509-cert/src/builder.rs (hunk 2 above)
Comment (lumag) [reply x509-cert/src/builder.rs:170]: @baloo it is reasoned in the commit message. Please review by the commit.

**Comment by lumag** [comment]
@tarcieri I think the option 1 sounds better. In fact I think there is option 3: switch to `SignerMut` and try wrapping `RadomizedSigner` in the `SignerMut`, but I think it is worse than 1.

**Comment by tarcieri** [comment]
I think we could also potentially do both. Having a `Signer` impl on PSS seems nice when `OsRng` is available.

**Comment by baloo** [comment]
> @baloo another question: would it be useful to accept `PK: EncodePublicKey` instead of `SPKIOwned`?

That would save a round trip to der serialization, and it's more consistent with the signer. Up to you.

**Comment by baloo** [comment]
(the rest of the changes are okay with me, I've pulled the changes over to `yubikey.rs` to see and it works as expected)

**Comment by lumag** [comment]
@baloo I stumled upon another issue there. The builder will use signer's `signatureAlgorithm` for the `TbsCertificate.signature`, which is obviously incorrect (see the example from the leaf_certificate test). So, we will have to pass `PublicKey + DynSignatureAlgorithmIdentifier` to the `CertificateBuilder::new()`. I'm working on a patch to fix that.

**Comment by lumag** [comment]
for the reference:
```
---- leaf_certificate stdout ----
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 42 (0x2a)
        Signature Algorithm: ecdsa-with-SHA256
        Issuer: CN = World domination corporation, O = World domination Inc, C = US
        Validity
            Not Before: Apr 17 23:41:31 2023 GMT
            Not After : Apr 17 23:41:36 2023 GMT
        Subject: CN = service.domination.world
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:b6:c4:2c:51:5f:10:a6:aa:f2:82:c6:3e:db:e2:
                    42:43:a1:70:f3:fa:26:33:bd:48:33:63:7f:47:ca:
                    4f:6f:36:e0:3a:5d:29:ef:c3:19:1a:c8:0f:39:0d:
                    87:4b:39:e3:0f:41:4f:ce:c1:fc:a0:ed:81:e5:47:
                    ed:c2:cd:38:2c:76:f6:1c:90:18:97:3d:b9:fa:53:
                    79:72:a7:c7:01:f6:b7:7e:09:82:df:c1:5f:c0:19:
                    27:ee:5e:7c:d9:4b:4f:59:9f:f0:70:13:a7:c8:28:
                    1b:df:22:dc:bc:9a:d7:ca:bb:7c:43:11:c9:82:f5:
                    8e:db:72:13:ad:45:58:b3:32:26:6d:74:3a:ed:81:
                    92:d1:88:4c:ad:b8:b1:47:39:a8:da:da:66:dc:97:
                    08:06:d9:c7:ac:45:0c:b1:3d:0d:7c:57:5f:b1:98:
                    53:4f:c6:1b:c4:1b:c0:f0:57:4e:0e:01:30:c7:bb:
                    bf:bd:fd:c9:f6:a6:e2:e3:e2:af:f1:cb:ea:c8:9b:
                    a5:78:84:52:8d:55:cf:b0:83:27:a1:e8:c8:9f:4e:
                    00:3c:f2:88:8e:93:32:41:d9:d6:95:bc:bb:ac:dc:
                    90:b4:4e:3e:09:5f:a3:70:58:ea:25:b1:3f:5e:29:
                    5c:be:ac:6d:e8:38:ab:8c:50:af:61:e2:98:97:5b:
                    87:2f
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Subject Key Identifier: 
                73:85:97:8E:50:5F:C1:43:0F:2D:B5:A6:73:77:45:34:C8:1F:BC:44
            X509v3 Authority Key Identifier: 
                8D:07:D8:4F:5A:4E:48:6A:09:61:F8:F8:80:25:81:61:5B:30:97:80
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Key Usage: critical
                Digital Signature, Non Repudiation
    Signature Algorithm: ecdsa-with-SHA256
    Signature Value:
        e4:ae:bf:6f:a9:29:46:66:d3:51:ba:25:2d:07:f6:be:d4:83:
        32:f7:67:4c:4b:05:5e:4a:72:ca:89:f8:1f:a4:f1:75:6b:c2:
        94:b3:0e:ba:f5:a8:a3:8b:a0:7e:38:60:0a:5b:b2:43:55:17:
        63:53:2f:24:82:ef:4a:a3:ff:2e
```

The key is RSA, but the signatureAlgorithm is set to:
> Signature Algorithm: ecdsa-with-SHA256

**Comment by lumag** [comment]
After rereading RFC5280 I found that I was incorrect regarding the `TbsCertificate.signature` field. Please excuse me.

**Comment by tarcieri** [comment]
@lumag can you rebase?

**Comment by lumag** [comment]
@tarcieri Yes. I was going to push several additional changes, but probably I'd better move them to a separate PR as this one was reviewed by @baloo .

**Comment by lumag** [comment]
done

**Comment by tarcieri** [comment]
FWIW, this should allow RSASSA-PSS with the `Signer` trait: https://github.com/RustCrypto/RSA/pull/297

**Pull request merged**